partiql-parser/src/main/kotlin/org/partiql/parser/internal/PartiQLParserDefault.kt

```diff
@@ -557,29 +557,29 @@ internal class PartiQLParserDefault : PartiQLParser {
          *
          */
 
-        override fun visitAsIdent(ctx: GeneratedParser.AsIdentContext) = visitSymbolPrimitive(ctx.symbolPrimitive())
+        override fun visitAsIdent(ctx: GeneratedParser.AsIdentContext) = visitAs<Identifier.Symbol> (ctx.symbolPrimitive())
 
-        override fun visitAtIdent(ctx: GeneratedParser.AtIdentContext) = visitSymbolPrimitive(ctx.symbolPrimitive())
+        override fun visitAtIdent(ctx: GeneratedParser.AtIdentContext) = visitAs<Identifier.Symbol> (ctx.symbolPrimitive())
 
-        override fun visitByIdent(ctx: GeneratedParser.ByIdentContext) = visitSymbolPrimitive(ctx.symbolPrimitive())
+        override fun visitByIdent(ctx: GeneratedParser.ByIdentContext) = visitAs<Identifier.Symbol> (ctx.symbolPrimitive())
 
-        override fun visitSymbolPrimitive(ctx: GeneratedParser.SymbolPrimitiveContext) = translate(ctx) {
-            when (ctx.ident.type) {
-                GeneratedParser.IDENTIFIER_QUOTED -> identifierSymbol(
-                    ctx.IDENTIFIER_QUOTED().getStringValue(),
-                    Identifier.CaseSensitivity.SENSITIVE,
-                )
-                GeneratedParser.IDENTIFIER -> identifierSymbol(
-                    ctx.IDENTIFIER().getStringValue(),
-                    Identifier.CaseSensitivity.INSENSITIVE,
-                )
-                else -> throw error(ctx, "Invalid symbol reference.")
-            }
+        override fun visitQuotedIdentifier(ctx: org.partiql.parser.internal.antlr.PartiQLParser.QuotedIdentifierContext): Identifier.Symbol = translate(ctx) {
+            identifierSymbol(
+                ctx.IDENTIFIER_QUOTED().getStringValue(),
+                Identifier.CaseSensitivity.SENSITIVE
+            )
+        }
+
+        override fun visitUnquotedIdentifier(ctx: org.partiql.parser.internal.antlr.PartiQLParser.UnquotedIdentifierContext): Identifier.Symbol = translate(ctx) {
+            identifierSymbol(
+                ctx.text,
+                Identifier.CaseSensitivity.INSENSITIVE
+            )
         }
```

Comment:
Hmm, why do we need the `visitAs` when we have explicit return types? I wonder if adding the two alternatives caused this. I wonder if we should define our own `getIdentifier` that is more efficient than visitAs.

So it would be like
```kotlin
fun identifier(ctx: ParserRuleContext): Identifier = when (ctx) {
     is IdentifierQuotedContext -> identifier(ctx)
     is IdentifierUnquotedContext -> identifier(ctx)
}

fun identifier(ctx: IdentifierQuotedContext): Identifier = ....

fun identifier(ctx: IdentifierUquotedContext): Identifier.Symbol = ....
```

We also should take care where qualified names appear and make sure our grammar rules follow this.

Reply:
Like, I think we should be replacing `visitSymbolPrimitive` with `visitIdentifierUnquoted` and it shouldn't be anymore than a name change. Is there a reason this did not work?

Reply:
There wasn't a reason this did not work previously. I was just following the `visitAs<...> ()` pattern I often saw in the visitors. I'll update to use a `getIdentifier` (or the like) rather than `visitAs`.

Reply:
New commit [445be01](https://github.com/partiql/partiql-lang-kotlin/pull/1457/commits/445be01db417657469c8b833b93c6b7b4b63965c) uses the private `visitSymbolPrimitive` function which calls the corresponding `visitIdentifierQuoted` and `visitIdentifierUnquoted`. Also changed some existing usages of `visitAs<Identifier.Symbol>` to use `visitSymbolPrimitive`.